aws/retry/retryable_error.go

```diff
@@ -0,0 +1,168 @@
+
+// IsErrorRetryable returns if the error is retryable if any of the checks in
+// the list return a value other than unknown.
+func (r IsErrorRetryables) IsErrorRetryable(err error) aws.Ternary {
+	for _, re := range r {
+		if v := re.IsErrorRetryable(err); v != aws.UnknownTernary {
+			return v
+		}
+	}
+	return aws.UnknownTernary
+}
+
+// IsErrorRetryableFunc wraps a function with the IsErrorRetryable interface.
+type IsErrorRetryableFunc func(error) aws.Ternary
+
+// IsErrorRetryable returns if the error is retryable.
+func (fn IsErrorRetryableFunc) IsErrorRetryable(err error) aws.Ternary {
+	return fn(err)
+}
+
+// RetryableOperationError is an IsErrorRetryable implementation which uses the
+// optional interface Retryable on the error value to determine if the error is
+// retryable.
+type RetryableOperationError struct{}
+
+// IsErrorRetryable returns if the error is retryable if it satisfies the
+// Retryable interface, and returns if the attempt should be retried.
+func (RetryableOperationError) IsErrorRetryable(err error) aws.Ternary {
+	var v interface{ RetryableError() bool }
+
+	if !errors.As(err, &v) {
+		return aws.UnknownTernary
+	}
+
+	return aws.BoolTernary(v.RetryableError())
+}
+
+// NoRetryCanceledError detects if the error was an request canceled error and
+// returns if so.
+type NoRetryCanceledError struct{}
+
+// IsErrorRetryable returns the error is not retryable if the request was
+// canceled.
+func (NoRetryCanceledError) IsErrorRetryable(err error) aws.Ternary {
+	var v interface{ CanceledError() bool }
+
+	if !errors.As(err, &v) {
+		return aws.UnknownTernary
+	}
+
+	if v.CanceledError() {
+		return aws.FalseTernary
+	}
+	return aws.UnknownTernary
+}
+
+// RetryableConnectionError determines if the underlying error is an HTTP
+// connection and returns if it should be retried.
+type RetryableConnectionError struct{}
+
+// IsErrorRetryable returns if the error is caused by and HTTP connection
+// error, and should be retried.
+func (RetryableConnectionError) IsErrorRetryable(err error) aws.Ternary {
+	if err == nil {
+		return aws.UnknownTernary
+	}
+	var conErrVal bool
+
+	var conErr interface{ ConnectionError() bool }
+	var tempErr interface{ Temporary() bool }
+	var urlErr *url.Error
+	var netOpErr *net.OpError
+
+	switch {
+	case errors.As(err, &conErr):
+		conErrVal = conErr.ConnectionError()
+
+	case errors.As(err, &urlErr):
+		conErrVal = strings.Contains(urlErr.Error(), "connection refused")
+
+	case errors.As(err, &netOpErr):
+		conErrVal = netOpErr.Op == "dial"
+		if !conErrVal {
+			conErrVal = strings.Contains(netOpErr.Error(), "connection reset")
+		}
+
+	case errors.As(err, &tempErr):
+		// url.Error and net.OpError implement Temporary but have different
+		// expectations for when an error is or is not temporary.
+		conErrVal = tempErr.Temporary()
+
+	default:
+		if strings.Contains(err.Error(), "connection reset") {
+			return aws.TrueTernary
+		}
+		return aws.UnknownTernary
```

Comment:
Is there cases where this happens and it is not a net.OpErrror?

Reply:
yeah this is more to cover the bases, where anything that contains connection reset is considered a connection error. This is what the old logic did, but sprinkled in a few places through the checks. With this update the logic is centralized here. I have a few updates to this RetryConnectionError that I'll publish shortly.